### ISSUE TITLE
Dump/Extract File

### ISSUE THREAD

**Issue by dscdo** [libvmi/libvmi#704]
Hi.
With libvmi extract specific file from target vm is it possible?

**Comment by smaresca** [comment]
LibVMI itself does not offer provide an abstraction to extract files, but functions in its API could be used to achieve that result. 

Example methods:
1) Assuming the file is fully or partly memory-resident a process of the VM, identify its base address+size, and use the vmi_read_* family of functions to copy that data into your observing program
2) Intercept functions used to read/write files as they are called within a running VM, copying the data to your observing program
3) Inject a process to read on your behalf and extract the data so that it's accessible via #1 or #2
4) Temporarily use an executing 'donor' process to call relevant APIs or map the file so that it's accessible via #1 or #2
5) etc

Bottom line prerequisite for extracting a file: its data must be available in memory at time of introspection. LibVMI has the same requirements as Volatility, Rekall, and other static memory analysis tools in this regard.

Take a look at drakvuf for example implementations for file extraction

**Comment by tklengyel** [comment]
You are likely looking for `libguestfs`: http://libguestfs.org
